### PR TITLE
[Blob Tag] Phase I - Implement SetTag, GetTag on differenet blob APIs

### PR DESCRIPTION
**Pull request by blueww**
Thanks for contribution! Please go through following checklist before sending PR.
This is a PR for Blob tag Phase I change:

Implement set/get BlobTag on API: GetBlobTag, SetBlobTag, GetBlob, GetBlobProperties, PUTBlob (Block/page/append), ListBlob, PutBlockList, copyFromURL, startCopyFromURL.

https://github.com/Azure/Azurite/issues/1871
https://github.com/Azure/Azurite/issues/1872

### PR Branch Destination

- For Azurite V3, please send PR to `main` branch.
- For legacy Azurite V2, please send PR to `legacy-dev` branch.

### Always Add Test Cases

Make sure test cases are added to cover the code change.

### Add Change Log

Add change log for the code change in `Upcoming Release` section in `ChangeLog.md`.

### Development Guideline

Please go to CONTRIBUTION.md for steps about setting up development environment and recommended Visual Studio Code extensions.

### CODE REVIEW

File: src/blob/persistence/LokiBlobMetadataStore.ts [hunk 1]
```diff
@@ -3299,6 +3306,103 @@ export default class LokiBlobMetadataStore
...
+      account,
+      container,
+      blob,
+      snapshot,
+      context,
+      false,
+      true
+    );
+
+    validateWriteConditions(context, modifiedAccessConditions, doc);
```
Comment (XiaoningLiu) [inline src/blob/persistence/LokiBlobMetadataStore.ts:3344]: Does set tag support access conditions?

File: src/blob/persistence/SqlBlobMetadataStore.ts [hunk 2]
```diff
@@ -289,6 +290,9 @@ export default class SqlBlobMetadataStore implements IBlobMetadataStore {
         },
         metadata: {
           type: "VARCHAR(2047)"
+        },
+        blobTags: {
+          type: "VARCHAR(2047)"
```
Comment (XiaoningLiu) [inline src/blob/persistence/SqlBlobMetadataStore.ts:295]: What's the maxmium size of blob tags for single blob?

File: src/blob/utils/utils.ts [hunk 3]
```diff
@@ -164,3 +165,79 @@ export function getUserDelegationKeyValue(
...
+  if (blobTagsString === '' || blobTagsString === undefined)
+  {
+    return undefined;
+  }
+  let blobTags:BlobTag[] = [];
+  const rawTags = blobTagsString.split("&");
+  rawTags.forEach((rawTag)=>{
+    const tagpair = rawTag.split("=");
+    blobTags.push({
+      key: decodeURIComponent(tagpair[0]),
```
Comment (XiaoningLiu) [inline src/blob/utils/utils.ts:185]: does server has same behavior on URI decode against tags?

**Review by XiaoningLiu — APPROVED** [review]
Approved with comments.

**Review by XiaoningLiu — APPROVED** [review]
Approved with comments.

File: src/blob/persistence/LokiBlobMetadataStore.ts (hunk 1 above)
Comment (blueww) [reply src/blob/persistence/LokiBlobMetadataStore.ts:3344]: Good catch, it doesn't support access condition, except x-ms-if-tags.

File: src/blob/persistence/SqlBlobMetadataStore.ts (hunk 2 above)
Comment (blueww) [reply src/blob/persistence/SqlBlobMetadataStore.ts:295]: Per https://learn.microsoft.com/en-us/rest/api/storageservices/set-blob-tags?tabs=azure-ad#request-body , tag count <= 10, key length <= 128, value length  <=256
Add up the other chars like keyword, blanket, for 10 tags, will need < 250 chars.
So the total should be 4090, I change it to 4096

File: src/blob/utils/utils.ts (hunk 3 above)
Comment (blueww) [reply src/blob/utils/utils.ts:185]: I have tested all allowed special chars that allowed in blob tag in test case, all works well.
The input chars is same as output chars.

When we get the input tag here, the special char is encoded, so we need decode it.

File: src/blob/persistence/LokiBlobMetadataStore.ts [hunk 4]
```diff
@@ -3299,6 +3306,103 @@ export default class LokiBlobMetadataStore
...
+   * @returns {Promise<BlobTags | undefined>}
+   * @memberof LokiBlobMetadataStore
+   */
+  public async getBlobTag(
+    context: Context,
+    account: string,
+    container: string,
+    blob: string,
+    snapshot: string = "",
+    leaseAccessConditions: Models.LeaseAccessConditions | undefined,
```
Comment (EmmaZhu) [inline src/blob/persistence/LokiBlobMetadataStore.ts:3378]: Does get tags operations require lease ID for a leased blob?

File: src/blob/utils/utils.ts [hunk 5]
```diff
@@ -164,3 +165,81 @@ export function getUserDelegationKeyValue(
...
+    if (ContainsInvalidTagCharacter(tag.value)) {
+      throw StorageErrorFactory.getInvalidTag(contextID);
+    }
+  });
+}
+
+function ContainsInvalidTagCharacter(s: string): boolean{
+  for (let c of s)
+  {
+    if (!(c >= 'a' && c <= 'z' ||
```
Comment (EmmaZhu) [inline src/blob/utils/utils.ts:229]: Add embrace around (c >= 'a' && c <= 'z') to make the logic clearer.

File: tests/blob/apis/blob.test.ts [hunk 6]
```diff
@@ -1327,6 +1328,472 @@ describe("BlobAPIs", () => {
...
+    const guid = "ca761232ed4211cebacd00aa0057b223";
+    const leaseClient = blockBlobClient.getBlobLeaseClient(guid);
+    await leaseClient.acquireLease(-1);
+
+    const tags = {
+      tag1: "val1",
+      tag2: "val2",
+    };
+    await blockBlobClient.setTags(tags, { conditions: { leaseId: leaseClient.leaseId } });
+    const response = await blockBlobClient.getTags({
```
Comment (EmmaZhu) [inline tests/blob/apis/blob.test.ts:1769]: Should getTags still work without lease ID?

**Review by EmmaZhu — APPROVED** [review]

File: src/blob/persistence/LokiBlobMetadataStore.ts (hunk 4 above)
Comment (blueww) [reply src/blob/persistence/LokiBlobMetadataStore.ts:3376]: Per the rest API doc, it does: https://learn.microsoft.com/en-us/rest/api/storageservices/get-blob-tags?tabs=azure-ad#request-headers

File: src/blob/utils/utils.ts (hunk 5 above)
Comment (blueww) [reply src/blob/utils/utils.ts:229]: As we discussed, we agree with current code.

**Pull request merged**